### ISSUE TITLE
Similarity and difference between `evaluate`, `fit''s validation option, `test_on_batch` and `train_on_batch`

### ISSUE THREAD

**Issue by fluency03** [fchollet/keras#2078]
I am willing to clarify my confusion regarding these methods `evaluate`, `fit', `test_on_batch` and `train_on_batch`. 

1. One confusion is that: in the `fit`, there is option that you can set portion of the training data as the testing data.  Is that the same, if I just choose same percentage of my data and put them into `evaluate`?  And the only difference is, as how I understand: in `fit`, test set are chosen randomly, whereas in `evaluate`, the test set are just the data you give. 

2. There are also methods like `test_on_batch` and `train_on_batch`. The difference between them and `fit` and `evaluate` is that: in `test_on_batch` and `train_on_batch`, you only test or train on one batch, but in `fit` and `evaluate`, you do this on plenty of batches. 

Thanks.

**Comment by EderSantana** [comment]
> One confusion is that: in the fit, there is option that you can set portion of the training data as the testing data. Is that the same, if I just choose same percentage of my data and put them into evaluate? And the only difference is, as how I understand: in fit, test set are chosen randomly, whereas in evaluate, the test set are just the data you give.

When we compile a Keras model, there are two functions as output. `_train` and `_test`. `train_on_batch` and `test_on_batch` are wrappers around those compiled functions, this is to provide a friendly API. 

The difference between `_train` and `_test` is for some types of layer, like Dropout that behave differently in train and test. Also, when you call `_train` it will do a gradient update in your weights. `fit` calls `_train` on the training dataset and `_test` on the validation dataset. `evaluate` will also call `_test` on the evaluation dataset.

The reason why `_test` is called inside `fit` is because cross-validation and early stopping is really important for neural nets. Also, there are two ways to define your validation dataset. First you can simply say the validation/training split ratio, the other is defining the `validation_data` directly.


> There are also methods like test_on_batch and train_on_batch. The difference between them and fit and evaluate is that: in test_on_batch and train_on_batch, you only test or train on one batch, but in fit and evaluate, you do this on plenty of batches.

yes!

**Comment by fluency03** [comment]
@EderSantana Thanks for your answers. 

I am wondering, there is validation part in the `fit`, where you can do the cross validation, right? So why have `evaluate`? From my understanding, `evaluate` is also for validating right?

**Comment by EderSantana** [comment]
`evaluate` is for "testing". The difference is a neural net jargon. We call validation the development "test". Something we use as we train the network to help us choose hyperparameters and early-stopping. 

Testing as what you do only after your neural net is fully trained. You shouldn't use information in the test set to fine-tune your network. So think about `evaluate` as something you will do only after you are done training in a dataset that was not used as your `validation_data`. 

Also, Keras goes beyond academic uses. Some people want to test their trained neural network in production (or Kaggle) without needing to retrain it every time.

**Comment by fluency03** [comment]
@EderSantana thanks a lot. it's clear to me.
